api/src/opentrons/protocol_api/core/engine/instrument.py

```diff
@@ -843,3 +844,27 @@ def retract(self) -> None:
         """Retract this instrument to the top of the gantry."""
         z_axis = self._engine_client.state.pipettes.get_z_axis(self._pipette_id)
         self._engine_client.execute_command(cmd.HomeParams(axes=[z_axis]))
+
+    def find_liquid_level(self, well_core: WellCore) -> float:
+        labware_id = well_core.labware_id
+        well_name = well_core.get_name()
+        well_location = WellLocation(
+            origin=WellOrigin.TOP, offset=WellOffset(x=0, y=0, z=0)
+        )
+        result = self._engine_client.execute_command_without_recovery(
```

Comment:
```suggestion
        result = self._engine_client.execute_command(
```
we definitely want to recover from this command

Reply:
I think the problem here is that one of this method's callers, `InstrumentContext.get_liquid_height()`, is returning the probed height. And if the `liquidProbe` command fails and we let the frontend drive error recovery, there *isn't* a probed height in the general case. The client can choose to skip to the next command without retrying the failed `liquidProbe`.

We can certainly solve this for *`InstrumentContext.require_liquid()`*. Just, yeah, call `execute_command()` (with recovery) and discard the `liquidProbe` result. A successful `liquidProbe`, or a failed-but-recovered `liquidProbe`, will return from `execute_command()`. A failed-and-not-recovered `liquidProbe` will raise from `execute_command()`, and we should let that propagate up.

But for `InstrumentContext.get_liquid_height()` and `InstrumentContext.detect_liquid_presence()`, it's unclear to me what we would even want to happen.

Reply:
Recap of in-person discussion. Looks like you've already done some of this by the time I'm posting this comment.

## `InstrumentContext.require_liquid_presence()`

We want this method to give the client the opportunity for error recovery, but raise an exception if that ultimately fails. That's what `SyncClient.execute_command()`, so under the hood, this should be implemented using that. If that raises an exception, we should just let it propagate.

## `InstrumentContext.detect_liquid_presence()`

We want this method to straight-up return `True`/`False` *without* giving the client the opportunity for error recovery. So, under the hood:

* We should implement it via `SyncClient.execute_command_without_recovery()`.
* If the probe detects no liquid, `execute_command_without_recovery()` will raise that as an exception. We should look for that and translate it into a `False` return.

  We can catch `ProtocolCommandFailedError`, and then check if `exception.original_error` is an instance of [`LiquidNotFoundError`](https://github.com/Opentrons/opentrons/blob/6ba97194fc65349a91a24b41b5c244c69bb8e6ed/api/src/opentrons/protocol_engine/commands/pipetting_common.py#L150). If so, we should translate it into a `False` return. Otherwise, we should re-raise the exception to let it propagate.

I think this will expose a bug where the robot will mistakenly go into error recovery mode (because "the command failed") *and* try to keep running the protocol (because we're catching the exception, translating it to `False`, and letting the protocol continue). We can take that on in a separate PR: EXEC-598.

## `InstrumentContext.measure_liquid_height()`

We want to keep this internal and experimental for now and keep our options open for the future. So:

* We should say in its docstring that it's internal and experimental, and exclude it from the public docs.opentrons.com website. [Here is an example of how to exclude it from docs.opentrons.com.](https://github.com/Opentrons/opentrons/blob/6ba97194fc65349a91a24b41b5c244c69bb8e6ed/api/src/opentrons/protocol_api/disposal_locations.py#L207-L214)
* We should implement this via `SyncClient.execute_command_without_recovery()`. So if no liquid is detected, it will raise an exception. We should just let that propagate.

The reason we want to keep this internal and experimental is that we haven't really fleshed out how to support error recovery on PAPI methods that return results. Does the UI flow need to have a step to retry the failed probe? Should we just raise some documented exception for the protocol author to catch and deal with how they please?